network/framework/src/protocols/wire/messaging/v1/metadata.rs

```diff
@@ -0,0 +1,279 @@
+    }
+
+    /// Returns a reference to the RPC response
+    pub fn rpc_response(&self) -> &RpcResponse {
+        &self.response
+    }
+}
+
+/// A simple enum to track the message send latency metric
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum MessageSendLatency {
+    ApplicationToWire,
+    WireSend,
+}
+
+impl MessageSendLatency {
+    pub fn get_label(&self) -> &'static str {
+        match self {
+            MessageSendLatency::ApplicationToWire => "ApplicationToWire",
+            MessageSendLatency::WireSend => "WireSend",
+        }
+    }
+}
+
+/// A simple enum to track the message send type
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum MessageSendType {
+    DirectSend,  // A direct send message to another peer
+    RpcRequest,  // An RPC request to another peer
+    RpcResponse, // An RPC response for a request sent by another peer
+}
+
+impl MessageSendType {
+    pub fn get_label(&self) -> &'static str {
+        match self {
+            MessageSendType::DirectSend => "DirectSend",
+            MessageSendType::RpcRequest => "RpcRequest",
+            MessageSendType::RpcResponse => "RpcResponse",
+        }
+    }
+}
+
+/// A simple enum to track the message stream type
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum MessageStreamType {
+    NonStreamedMessage,      // A non-streamed message that fits into a single chunk
+    StreamedMessageHead,     // The head (first fragment) of a streamed message
+    StreamedMessageFragment, // A fragment of a streamed message (not the head or tail)
+    StreamedMessageTail,     // The tail (last fragment) of a streamed message
+}
+
+impl MessageStreamType {
+    pub fn get_label(&self) -> &'static str {
+        match self {
+            MessageStreamType::NonStreamedMessage => "NonStreamedMessage",
+            MessageStreamType::StreamedMessageHead => "StreamedMessageHead",
+            MessageStreamType::StreamedMessageFragment => "StreamedMessageFragment",
+            MessageStreamType::StreamedMessageTail => "StreamedMessageTail",
+        }
+    }
+}
+
+/// A struct holding metadata about each message.
+/// Note: this is not sent along the wire, it is only used internally.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct MessageMetadata {
+    /// The network ID for the message
+    network_id: NetworkId,
+
+    /// The protocol ID for the message. This may not always be
+    /// known (e.g., when failing to deserialize a message).
+    protocol_id: Option<ProtocolId>,
+
+    /// The type of message being sent
+    message_send_type: MessageSendType,
+
+    /// The stream type of the message being sent
+    message_stream_type: MessageStreamType,
+
+    /// The time at which the message was sent by the application
+    application_send_time: Option<SystemTime>,
+
+    /// The time at which the message started being sent over the wire
+    wire_send_start_time: Option<SystemTime>,
+}
+
+impl MessageMetadata {
```

Comment:
what if the message is dropped somewhere before it's sent? Would it be useful to add some sort of Drop impl to observe this? This can happen for Rpcs often I believe, if the rx oneshot is dropped by the application.

Reply:
Yeah, currently, if a message is dropped before it's sent, we don't log any latency metrics via this mechanism. We could if we wanted (e.g., via the `Drop impl`), but I'm not sure exactly what we'd want to track? 🤔 Were you thinking the number of dropped sends, or something else? 😄

Reply:
Yes, the dropped count would be helpful. I think we have channel dropped counters, but I am not sure if we explicitly decide to drop for some reason. Basically, I am thinking this as a packet loss metric within our stack.

Reply:
Gotcha, makes sense 😄 If it's okay with you, I'll do this in a separate PR. That way, I can: (i) avoid making this one larger; and (ii) also do it for the receiver side metrics (so we can see the drop rate across the entire stack).

Reply:
Sure thing.